staging/src/k8s.io/kms/encryption/service_test.go

```diff
@@ -361,29 +437,331 @@ func TestDecryptError(t *testing.T) {
+	if remoteKMS.EncryptCallCount() != 1 {
+		t.Fatalf("Encrypt() remoteKMS.EncryptCallCount() = %v, want %v", remoteKMS.EncryptCallCount(), 1)
+	}
+
+	var wg sync.WaitGroup
+	for i := 0; i < 6; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			resp, err := localKEKService.Encrypt(ctx, "test-uid", []byte(rand.String(32)))
+			if err != nil {
+				t.Fatalf("Encrypt() error = %v", err)
+			}
+			if v, ok := resp.Annotations[referenceKEKAnnotationKey]; !ok || !bytes.Equal(v, encLocalKEK) {
+				t.Fatalf("Encrypt() annotations = %v, want %v", resp.Annotations, encLocalKEK)
+			}
+			record.Store(resp, nil)
+		}()
+	}
+	wg.Wait()
+
+	rotated := false
+	// wait for the local KEK to be rotated
+	wait.PollImmediate(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+		// local KEK must have been rotated after 5 usages
+		lk = localKEKService.localKEKTracker.Load()
+		rotated = !bytes.Equal(lk.encKEK, encLocalKEK)
+		return rotated, nil
+	})
+	if !rotated {
+		t.Fatalf("local KEK must have been rotated")
+	}
+	if remoteKMS.EncryptCallCount() != 2 {
+		t.Fatalf("Encrypt() remoteKMS.EncryptCallCount() = %v, want %v", remoteKMS.EncryptCallCount(), 2)
+	}
+
+	// new local KEK must be used for encryption
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			resp, err := localKEKService.Encrypt(ctx, "test-uid", []byte(rand.String(32)))
+			if err != nil {
+				t.Fatalf("Encrypt() error = %v", err)
+			}
+			if v, ok := resp.Annotations[referenceKEKAnnotationKey]; !ok || !bytes.Equal(v, lk.encKEK) {
+				t.Fatalf("Encrypt() annotations = %v, want %v", resp.Annotations, lk.encKEK)
+			}
+			record.Store(resp, nil)
+		}()
+	}
+	wg.Wait()
+
+	// check we can decrypt data encrypted with the old and new local KEKs
+	record.Range(func(key, _ any) bool {
+		k := key.(*service.EncryptResponse)
+		decryptRequest := &service.DecryptRequest{
+			Ciphertext:  k.Ciphertext,
+			Annotations: k.Annotations,
+			KeyID:       k.KeyID,
+		}
+		if _, err := localKEKService.Decrypt(ctx, "test-uid", decryptRequest); err != nil {
+			t.Fatalf("Decrypt() error = %v", err)
+		}
+		return true
+	})
+
+	// Out of the 11 calls to Decrypt:
+	// - 5 should be using the current local KEK
+	// - 1 out of the 6 should generate a decrypt call to the remote KMS as the local KEK not in cache
+	// - 5 out of the 6 should use the cached local KEK after 1st decrypt call to the remote KMS
+	if remoteKMS.DecryptCallCount() != 1 {
+		t.Fatalf("Decrypt() remoteKMS.DecryptCallCount() = %v, want %v", remoteKMS.DecryptCallCount(), 1)
+	}
+	if localKEKService.transformers.Len() != 1 {
+		t.Fatalf("Decrypt() localKEKService.transformers.Len() = %v, want %v", localKEKService.transformers.Len(), 1)
+	}
+}
+
+func TestRotationKeyExpiry(t *testing.T) {
+	ctx := testContext(t)
+
+	var record sync.Map
+
+	remoteKMS := &testRemoteService{keyID: "test-key-id"}
+	localKEKService := newLocalKEKService(ctx, remoteKMS, 10, 5, 2*time.Second, 100*time.Millisecond)
```

Comment:
I feel like this has a high chance of flaking in CI because it is totally for the 3 encrypt calls below to take longer than that time on a slow CI runner.

An alternative to the wait polls would be to just mock `time.Now` in tests so you can advance it as you desire.

Reply:
Added a `clock` and mocked that using fake clock in the tests. Even with the fake clock + step to advance the time, we'll need the wait polls to ensure the controller loop runs to rotate the key, so I've still kept that.